enhancements/20251105-pausable-node-upgrades.md

```diff
@@ -0,0 +1,158 @@
+
+### User Experience In Detail
+
+The user can set the node upgrade **mode** to be `auto` or `manual`. The default value is `auto`, which suggests the node upgrade behavior we're all familiar with: nodes are upgraded one by one autonomously. If the mode is set to `manual`, the user can further set the pausability of upgrades for specific nodes in the **pauseNodes** field. Nodes not in the **pauseNodes** field will still undergo node upgrades automatically; if no nodes are specified in **pauseNodes**, all nodes will be paused before entering the pre-draining state, provided the mode is `manual`. When the **mode** is set to `auto` or not being set at all, the values in the **pauseNodes** are completely ignored.
+
+For instance, the following setting will pause upgrades for nodes `node-2`, `node-5`, and `node-9`. Other nodes in the cluster will still be upgraded autonomously.
+
+```json
+{
+  "nodeUpgradeOption": {
+    "strategy": {
+      "mode": "manual",
+      "pauseNodes": [
+        "node-2",
+        "node-5",
+        "node-9"
+      ]
+    }
+  }
+}
+```
+
+Another example, the following setting will pause all the nodes in the cluster before they enters the pre-draining state.
+
+```json
+{
+  "nodeUpgradeOption": {
+    "strategy": {
+      "mode": "manual"
+    }
+  }
+}
+```
+
+When the user triggers an upgrade for the cluster by clicking the "Upgrade" button on the dashboard UI or manually creating the Upgrade custom resource, the relevant setting will be automatically applied. Note that it will not take effect if the user configures the setting after the upgrade has started. During the node upgrade phase, the user can observe each node's upgrade progress in the Upgrade custom resource and in the upgrade dialog.
+
+After the user completes the tasks and is ready to continue with the node upgrade, they can remove the annotation from the node or click the "Resume" button on the upgrade dialog to proceed.
+
+### API changes
+
+There are no actual API changes involved.
+
+## Design
+
+### Implementation Overview
+
+The implementation includes two major parts: a set of new setting fields and the node annotation manipulation. Users configure the `nodeUpgradeOption` fields in the `upgrade-config` setting to specify their node upgrade preferences. The upgrade controller then annotates the nodes marked for pause according to the setting when the upgrade is initiated.
+
+During the node upgrade phase, the secret controller reconciles the machine-plan secrets. When the time comes (when the machine-plan secret for a node is annotated as eligible to upgrade), it checks whether the node has the annotation before creating the pre-drain job. When the annotation is removed from the node, the secret controller is triggered again and reconciles the machine-plan secrets to create the corresponding pre-drain job, thus continuing the original node upgrade flow.
```

Comment:
Does Harvester or Rancher have related time-out, if this step does not move ahead?

Reply:
Timeout is not required for now.

We can add this to the document, controller waits without timeout. Meanwhile, some Harvester webhooks might disallow user's operation inbetween the upgrade.